java/java.file.launcher/src/org/netbeans/modules/java/file/launcher/queries/MultiSourceRootProvider.java

```diff
@@ -0,0 +1,300 @@
+
+    public static boolean DISABLE_MULTI_SOURCE_ROOT = false;
+
+    //TODO: the cache will probably be never cleared, as the ClassPath/value refers to the key(?)
+    private Map<FileObject, ClassPath> file2SourceCP = new WeakHashMap<>();
+    private Map<FileObject, ClassPath> root2SourceCP = new WeakHashMap<>();
+    private Map<FileObject, ClassPath> file2AllPath = new WeakHashMap<>();
+    private Map<FileObject, ClassPath> file2ClassPath = new WeakHashMap<>();
+    private Map<FileObject, ClassPath> file2ModulePath = new WeakHashMap<>();
+
+    @Override
+    public ClassPath findClassPath(FileObject file, String type) {
+        switch (type) {
+            case ClassPath.SOURCE: return getSourcePath(file);
+            case ClassPath.COMPILE:
+                return attributeBasedPath(file, file2AllPath, "-classpath", "-cp", "--class-path", "--module-path", "-p");
+            case JavaClassPathConstants.MODULE_CLASS_PATH:
+                return attributeBasedPath(file, file2ClassPath, "-classpath", "-cp", "--class-path");
+            case JavaClassPathConstants.MODULE_COMPILE_PATH:
+                return attributeBasedPath(file, file2ModulePath, "--module-path", "-p");
+            case ClassPath.BOOT:
+            case JavaClassPathConstants.MODULE_BOOT_PATH:
+                return getBootPath(file);
+        }
+        return null;
+    }
+
+    private ClassPath getSourcePath(FileObject file) {
+        if (DISABLE_MULTI_SOURCE_ROOT) return null;
+        synchronized (this) {
+            //XXX: what happens if there's a Java file in user's home???
+            if (file.isData() && "text/x-java".equals(file.getMIMEType())) {
+                return file2SourceCP.computeIfAbsent(file, f -> {
+                    try {
+                        String content = new String(file.asBytes(), FileEncodingQuery.getEncoding(file));
+                        String packName = findPackage(content);
+                        FileObject root = file.getParent();
+
+                        if (packName != null) {
+                            List<String> packageParts = Arrays.asList(packName.split("\\."));
+
+                            Collections.reverse(packageParts);
+
+                            for (String packagePart : packageParts) {
+                                if (!root.getNameExt().equalsIgnoreCase(packagePart)) {
+                                    //ignore files outside of proper package structure,
+                                    //those may too easily lead to using a too general
+                                    //directory as a root, leading to too much indexing:
+                                    return null;
+                                }
+                                root = root.getParent();
+                            }
+                        }
+
+                        return root2SourceCP.computeIfAbsent(root, r -> { //XXX: weak....
```

Comment:
This is unfortunately quite tricky. The CP must be registered in the `GlobalPathRegistry` (so that refactoring would work), so it will never be freed anyway. The current state is that these objects is basically perpetual (until the VM is closed, of course). I am afraid I don't have a good idea how to improve that. Probably the most realistic thing that comes to my mind is to free roots that are "unused" for an extended period of time. Which has its own problems, of course.